dqvm/src/global-addresses.lisp

```diff
@@ -0,0 +1,181 @@
+  "Get the (global) block index of ADDRESS."
+  (when address
+    (assert (not (minusp address)))
+    (assert (< address (number-of-addresses global-addresses)))
+
+    (let ((initial-address (get-initial-address global-addresses address)))
+      (nth-value 0 (floor initial-address (block-size global-addresses))))))
+
+(defmethod get-rank-by-address ((global-addresses global-addresses) address)
+  "Find the MPI rank of ADDRESS using ADDRESSES.
+
+This method returns the rank where the address is located. If ADDRESS is NIL, the return value is also NIL."
+  (when address
+
+    (let ((number-of-processes (number-of-processes global-addresses))
+          (block-size (block-size global-addresses))
+          (blocks-per-process (blocks-per-process global-addresses))
+          (remainder-blocks (remainder-blocks global-addresses))
+          (number-of-blocks (number-of-blocks global-addresses))
+          (permutation (permutation global-addresses)))
+
+      (assert (not (minusp address)))
+      (assert (< address (number-of-addresses global-addresses)))
+
+      (let ((initial-address (get-initial-address global-addresses address)))
+        (if (< initial-address (* (- number-of-blocks remainder-blocks) block-size))
+            (nth-value 0 (floor initial-address (* blocks-per-process block-size)))
+            (mod (floor initial-address block-size) number-of-processes))))))
+
+(defmethod get-rank-by-block ((global-addresses global-addresses) block-index)
+  (let ((number-of-blocks (number-of-blocks global-addresses))
+        (number-of-processes (number-of-processes global-addresses))
+        (blocks-per-process (blocks-per-process global-addresses))
+        (remainder-blocks (remainder-blocks global-addresses)))
+
+    (assert (and (not (minusp block-index))
+                 (< block-index number-of-blocks)))
+
+    (if (< block-index (- number-of-blocks remainder-blocks))
+        (nth-value 0 (floor block-index blocks-per-process))
+        (- block-index (* blocks-per-process number-of-processes)))))
+
+(defmacro do-addresses-in-block ((var global-addresses block-index
+                                  &optional (result nil result-p))
+                                 &body body)
+  "Iterate VAR over addresses in BLOCK-INDEX. If RESULT is non-NIL, evaluate and RESULT before returning."
+  (alexandria:once-only (block-index)
+    (alexandria:once-only (global-addresses)
+      (alexandria:with-gensyms (block-size permutation initial-address)
+        `(loop :with ,block-size := (block-size ,global-addresses)
+               :with ,permutation := (permutation ,global-addresses)
+               :for ,initial-address :from (* ,block-index ,block-size)
+                 :below (* (1+ ,block-index) ,block-size)
+               :for ,var := (apply-qubit-permutation ,permutation ,initial-address)
+               :when (< ,block-index (number-of-blocks ,global-addresses))
+                 :do ,@body
```

Comment:
I recommend putting `body` in a `(progn ...)` so keywords don't get interpreted in the body.